lib/spack/spack/url_buildcache.py

```diff
@@ -0,0 +1,1262 @@
+    layout_version: int = CURRENT_BUILD_CACHE_LAYOUT_VERSION,
+) -> Type[URLBuildcacheEntry]:
+    """Given a layout version, return the class responsible for managing access
+    to buildcache entries of that version"""
+    if layout_version == 2:
+        return URLBuildcacheEntryV2
+    elif layout_version == 3:
+        return URLBuildcacheEntry
+    else:
+        raise UnknownBuildcacheLayoutError(
+            f"Cannot create buildcache class for unknown layout version {layout_version}"
+        )
+
+
+def check_mirrors_for_layout():
+    """Check configured mirrors, warning about any that are missing layout.json"""
+    if not config.get("config:check_mirrors_on_startup"):
+        return
+
+    cache_class = get_url_buildcache_class()
+    for mirror in MirrorCollection(binary=True).values():
+        if not cache_class.check_layout_json_exists(mirror.fetch_url):
+            msg = (
+                f"Configured mirror {mirror.name} is missing layout.json and has either \n"
+                "    never been pushed or is of an old layout version. Consider running \n"
+                "    'spack buildcache migrate' or rebuilding the specs in this mirror."
+            )
+            tty.warn(msg)
+
+
+def validate_checksum(file_path, checksum_algorithm, expected_checksum) -> None:
+    """Compute the checksum of the given file and raise if invalid"""
+    local_checksum = spack.util.crypto.checksum(hash_fun_for_algo(checksum_algorithm), file_path)
+
+    if local_checksum != expected_checksum:
+        size, contents = fsys.filesummary(file_path)
+        raise spack.error.NoChecksumException(
+            file_path, size, contents, checksum_algorithm, expected_checksum, local_checksum
+        )
+
+
+def _get_compressor(compression: str, writable: io.IOBase) -> io.IOBase:
+    if compression == "gzip":
+        return gzip.GzipFile(filename="", mode="wb", compresslevel=6, mtime=0, fileobj=writable)
+    elif compression == "none":
+        return writable
+    else:
+        raise BuildcacheEntryError(f"Unknown compression type: {compression}")
+
+
+@contextmanager
+def compression_writer(output_path: str, compression: str, checksum_algo: str):
+    """Create and return a writer capable of writing compressed data. Available
+    options for compression are "gzip" or "none", checksum_algo is used to pick
+    the checksum algorithm used by the ChecksumWriter.
+
+    Yields a tuple containing:
+        io.IOBase: writer that can compress (or not) as it writes
+        ChecksumWriter: provides checksum and length of written data
+    """
+    with open(output_path, "wb") as writer, ChecksumWriter(
+        fileobj=writer, algorithm=hash_fun_for_algo(checksum_algo)
+    ) as checksum_writer, closing(
+        _get_compressor(compression, checksum_writer)
+    ) as compress_writer:
+        yield compress_writer, checksum_writer
+
+
+def compress_and_write_spec(
+    output_path: str, spec_dict: dict, checksum_algo: str
```

Comment:
checksum_algo is unused. I would assert it is `sha256` for now since nothing else is implemented.

Reply:
Also the function is more generic than its name implies. Alternative name `compressed_json_from_dict` or something?